### PR TITLE
Fix WireGuard support detection

### PR DESCRIPTION
**Pull request by MichaIng**
Support was enabled automatically if a WireGuard package was found or could have been made available. But if the WireGuard kernel module is not available, it needs to be compiled. The required kernel headers are only reliably known for Raspberry Pi (Raspbian) and for amd64. This commit resolves the related issue where linux-image-amd64 was attempted to be installed on non-amd64 systems.

Additionally this commit resolves the issue that kernel headers were required and a DKMS build done, even if the module was builtin, when no WireGuard package was found.

The $NEED_WIREGUARD_REPO variable has been replaced with $AVAILABLE_WIREGUARD, which practically serves the same information and allows a simpler support check.

Fixes: #1180
_____________
Probably not the most beautiful solution, but it covers and explains all cases correctly. The large number of conditions and combinations always lead to large or ugly if-then-else blocks, I think 🤔.

### CODE REVIEW

**Comment by orazioedoardo** [comment]
The original idea was to make the script as generic as possible, even for future releases of Raspbian/Debian/Ubuntu, however in practice is indeed difficult. Putting that aside, the pull request is a bit hard to read, so a few questions:

https://github.com/pivpn/pivpn/pull/1201/files#diff-72c239a72654ca02e835e59af5796ff45057d3f8b9ba8653f6e64dda7761e2f7R495

This is the most generic: if module is built-in and the package is available, then wireguard can be installed. However I don't get the `|| $PLAT == 'Debian'`. On Debian there is a case when you have the module built-in but no package is available?

https://github.com/pivpn/pivpn/pull/1201/files#diff-72c239a72654ca02e835e59af5796ff45057d3f8b9ba8653f6e64dda7761e2f7R501

Do you need to check for `amd64`? PiVPN is expected to work on `armhf` and `arm64` versions of Ubuntu too.

**Comment by MichaIng** [comment]
> The original idea was to make the script as generic as possible, even for future releases of Raspbian/Debian/Ubuntu

It might become easier with Debian Bullseye, as the generic Debian kernels for ARM (linux-image-armmp and linux-image-arm64) are supporting an increasing number or ARM SoCs/SBCs, and then there is a known headers package for this as well. Until then, there is no generic way for ARM (aside of Raspberry Pi, assuming that the archive.raspberrypi.org repository is always used, not 100% true, but good enough). What you could do is checking if those Debian generic kernel packages are installed, and then assume it is used and the related headers package can be used to build the WireGuard module. Although... with Bullseye these kernels have the module builtin anyway 😄.

> the pull request is a bit hard to read

I totally agree, I was trying a minimal change to fix the current linux-headers-amd64 install on ARM: https://github.com/pivpn/pivpn/issues/1180#issuecomment-743445513
But that is not easy to understand as well and still builds the module in cases where it is builtin.

> However I don't get the `|| $PLAT == 'Debian'`. On Debian there is a case when you have the module built-in but no package is available?

Yes, on Debian and on Raspbian (but Raspbian support is covered with the condition below that line). Current stable Buster does not have the package available, Debian only via backports (not always enabled) and Raspbian does not have backports. But kernels with builtin WireGuard are more and more common, e.g. all Armbian mainline kernel images and for Raspberry Pi it's in the pipeline, developed on their "next" branch, an can be easily applied via `rpi-update`. To not block support or do DKMS builds where it is not required, this case needs to be covered.

First it was:
```
# Immediate support
[[ $WIREGUARD_BUILTIN == 1 && -n $AVAILABLE_WIREGUARD ]] ||
# Package can be made available via Bullseye repo
[[ $WIREGUARD_BUILTIN == 1 && ( $PLAT == 'Debian' || $PLAT == 'Raspbian' ) ]] ||
# Module can be build for all architectures supported by Raspbian
[[ $PLAT == 'Raspbian' ]] ||
```

Then I merged that into the two lines present now to reduce redundant conditions (builtin + Raspbian, Raspbian alone) and code lines. But probably the above is worth it to make it easier to understand?

**Comment by orazioedoardo** [comment]
> But probably the above is worth it to make it easier to understand? (I mean, update the pull request).

Yes, please. Easier to read is better than compact.

**Comment by MichaIng** [comment]
Okay, I hope reading through it from top to bottom (easiest => hardest case that is supported), makes more sense now.

File: auto_install/install.sh [hunk 1]
```diff
@@ -475,31 +475,33 @@ preconfigurePackages(){
...
+		# If the module is builtin and the package available, we only need to install wireguard-tools.
+		[[ $WIREGUARD_BUILTIN == 1 && -n $AVAILABLE_WIREGUARD ]] ||
+		# If the package is not available, on Debian and Raspbian we can add it via Bullseye repository.
+		[[ $WIREGUARD_BUILTIN == 1 && ( $PLAT == 'Debian' || $PLAT == 'Raspbian' ) ]] ||
+		# If the module is not builtin, on Raspbian we know the headers package: raspberrypi-kernel-headers
+		[[ $PLAT == 'Raspbian' ]] ||
+		# On Debian (and Ubuntu), we can only reliably assume the headers package for amd64: linux-image-amd64
+		[[ $PLAT == 'Debian' && DPKG_ARCH == 'amd64' ]] ||
+		# On Ubuntu, additionally the WireGuard package needs to be available, since we didn't test mixing Ubuntu repositories.
+		[[ $PLAT == 'Ubuntu' && DPKG_ARCH == 'amd64' && -n $AVAILABLE_WIREGUARD ]]
```
Comment (orazioedoardo) [inline auto_install/install.sh:502]: Missing dollar signs on `DPKG_ARCH`.

**Pull request merged**